### PR TITLE
separate preference

### PR DESCRIPTION
**Pull request by breeswish**

### CODE REVIEW

File: vj4/handler/home.py [hunk 1]
```diff
@@ -113,6 +113,20 @@ class UserChangemailWithCodeHandler(base.Handler):
...
+@app.route('/home/preference', 'home_preference')
+class HomePreferenceHandler(base.Handler):
+  @base.require_priv(builtin.PRIV_USER_PROFILE)
+  async def get(self):
+    self.render('home_preference.html')
+
+  @base.require_priv(builtin.PRIV_USER_PROFILE)
+  @base.post_argument
+  @base.require_csrf_token
+  async def post(self, **kwargs):
```
Comment (iceboy-sjtu) [inline vj4/handler/home.py]: Security hole

File: vj4/handler/home.py [hunk 2]
```diff
@@ -113,6 +113,20 @@ class UserChangemailWithCodeHandler(base.Handler):
...
+    await self.set_settings(**kwargs)
+    self.json_or_redirect(self.referer_or_main)
```
Comment (iceboy-sjtu) [inline vj4/handler/home.py]: Should redirect to self, not referer_or_main

File: vj4/locale/zh_CN.yaml [hunk 3]
```diff
@@ -47,6 +48,8 @@ Current Email: 当前电子邮件
 Current Location: 当前位置
 Current Password: 当前密码
 Default Code Language: 默认代码语言
+Default Code Template: 默认代码模板
+If leaved blank, the built-in template of the selected language will be used.: 若留空，则将使用对应语言的内置代码模板
```
Comment (iceboy-sjtu) [inline vj4/locale/zh_CN.yaml]: leaved -> left
对应 -> corresponding / selected -> 选择的

File: vj4/model/adaptor/setting.py [hunk 4]
```diff
@@ -10,41 +10,57 @@
 from vj4.util import locale
 
 Setting = functools.partial(
-  collections.namedtuple('Setting', ['family', 'key', 'factory', 'range', 'default',
+  collections.namedtuple('Setting', ['category', 'family', 'key', 'factory', 'range', 'default',
                                      'ui', 'name', 'desc']),
   range=None, default=None, ui='text', name='', desc='')
 
 # Setting keys should not duplicate with user keys or session keys.
 SETTINGS = [
```
Comment (iceboy-sjtu) [inline vj4/model/adaptor/setting.py]: PREFERENCE_SETTINGS = [...]
ACCOUNT_SETTINGS = [...]

File: vj4/model/adaptor/setting.py [hunk 5]
```diff
@@ -10,41 +10,57 @@
...
-  Setting('setting_function', 'send_code', int, range=constant.setting.FUNCTION_RANGE,
-          ui='select', name='Send Code after acceptance',
-          desc='If enabled, source code will be emailed to you after the submission is accepted.')]
+  Setting('account', 'setting_privacy', 'show_gender', int, range=constant.setting.PRIVACY_RANGE,
+          ui='select', name='Gender Visibility')]
 
 SETTINGS_BY_KEY = collections.OrderedDict(zip((s.key for s in SETTINGS), SETTINGS))
-SETTINGS_BY_FAMILY = collections.OrderedDict(
-  (f, list(l)) for f, l in itertools.groupby(SETTINGS, lambda s: s.family))
+SETTINGS_BY_CATEGORY = collections.OrderedDict(
```
Comment (iceboy-sjtu) [inline vj4/model/adaptor/setting.py]: Bad style - hard to read

File: vj4/model/adaptor/setting.py [hunk 6]
```diff
@@ -10,41 +10,57 @@
...
+  (c, collections.OrderedDict(
+    (f, list(fl)) for f, fl in itertools.groupby(list(cl), lambda s: s.family)))
+      for c, cl in itertools.groupby(SETTINGS, lambda s: s.category))
+
+def get_settings(category, get_setting):
```
Comment (iceboy-sjtu) [inline vj4/model/adaptor/setting.py]: No callback please

File: vj4/locale/zh_CN.yaml (hunk 3 above)
Comment (iceboy-sjtu) [inline vj4/locale/zh_CN.yaml]: . -> 。

**Comment by breeswish** [comment]
approved

File: vj4/handler/home.py (hunk 1 above)
Comment (breeswish) [inline vj4/handler/home.py]: now category is checked

File: vj4/model/adaptor/setting.py (hunk 4 above)
Comment (breeswish) [inline vj4/model/adaptor/setting.py]: this breaks get_setting

File: vj4/handler/home.py (hunk 1 above)
Comment (iceb0y) [inline vj4/handler/home.py]: still security hole - you should check it in the handler

File: vj4/handler/home.py [hunk 7]
```diff
@@ -117,13 +118,27 @@ class UserChangemailWithCodeHandler(base.Handler):
...
   async def get(self):
-    self.render('home_account.html')
+    self.render('home_settings.html', category='account', settings_to_form=self.settings_to_form)
 
   @base.require_priv(builtin.PRIV_USER_PROFILE)
   @base.post_argument
   @base.require_csrf_token
   async def post(self, **kwargs):
-    await self.set_settings(**kwargs)
+    await self.set_settings('account', **kwargs)
```
Comment (iceb0y) [inline vj4/handler/home.py]: same as below

File: vj4/handler/home.py [hunk 8]
```diff
@@ -117,13 +118,27 @@ class UserChangemailWithCodeHandler(base.Handler):
 class HomeAccountHandler(base.Handler):
   @base.require_priv(builtin.PRIV_USER_PROFILE)
   async def get(self):
-    self.render('home_account.html')
+    self.render('home_settings.html', category='account', settings_to_form=self.settings_to_form)
```
Comment (iceb0y) [inline vj4/handler/home.py]: settings_to_form=self.settings_to_form gives me bad code smell

File: vj4/handler/home.py [hunk 9]
```diff
@@ -117,13 +118,27 @@ class UserChangemailWithCodeHandler(base.Handler):
...
-    await self.set_settings(**kwargs)
+    await self.set_settings('account', **kwargs)
+    self.json_or_redirect(self.referer_or_main)
+
+
+@app.route('/home/preference', 'home_preference')
+class HomeAccountHandler(base.Handler):
+  @base.require_priv(builtin.PRIV_USER_PROFILE)
+  async def get(self):
+    self.render('home_settings.html', category='preference', settings_to_form=self.settings_to_form)
```
Comment (iceb0y) [inline vj4/handler/home.py]: Same as above

File: vj4/model/adaptor/setting.py (hunk 4 above)
Comment (iceb0y) [inline vj4/model/adaptor/setting.py]: SETTINGS = PREFERENCE_SETTINGS + ACCOUNT_SETTINGS

File: vj4/model/adaptor/setting.py [hunk 10]
```diff
@@ -61,11 +75,13 @@ def get_setting(self, key):
       return next(iter(setting.range))
     return setting.factory()
 
-  async def set_settings(self, **kwargs):
+  async def set_settings(self, category, **kwargs):
```
Comment (iceb0y) [inline vj4/model/adaptor/setting.py]: get_setting and set_settings should be symmetric in some way - should not accept category in set_settings but not in get_setting. I think you can just deal with individual settings and ignore category here.

File: vj4/model/adaptor/setting.py (hunk 4 above)
Comment (breeswish) [inline vj4/model/adaptor/setting.py]: good idea

**Comment by iceb0y** [comment]
approved

**Pull request merged**